sp-sim/src/sensors.rs

```diff
@@ -0,0 +1,218 @@
+                        value: Err(error.value),
+                        timestamp: error.timestamp,
+                    }
+                }
+            }
+            Self { last_data: Some(data), last_error: None } => SensorReading {
+                value: Ok(data.value),
+                timestamp: data.timestamp,
+            },
+            Self { last_data: None, last_error: Some(error) } => {
+                SensorReading {
+                    value: Err(error.value),
+                    timestamp: error.timestamp,
+                }
+            }
+            Self { last_data: None, last_error: None } => SensorReading {
+                value: Err(SensorDataMissing::DeviceNotPresent),
+                timestamp: 0, // TODO(eliza): what do?
+            },
+        }
+    }
+}
+
+impl Sensors {
+    pub(crate) fn from_component_configs<'a>(
+        cfgs: impl IntoIterator<Item = &'a SpComponentConfig>,
+    ) -> Self {
+        let mut sensors = Vec::new();
+        let mut by_component = HashMap::new();
+        for cfg in cfgs {
+            if cfg.sensors.is_empty() {
+                continue;
+            }
+            if !cfg
+                .capabilities
+                .contains(DeviceCapabilities::HAS_MEASUREMENT_CHANNELS)
+            {
+                panic!(
+                    "invalid component config: a device with sensors should \
+                     have the `HAS_MEASUREMENT_CHANNELS` capability:{cfg:#?}"
+                );
+            }
+
+            let mut ids = Vec::with_capacity(cfg.sensors.len());
+            for sensor in &cfg.sensors {
+                let sensor_id = sensors.len() as u32;
+                sensors.push(Sensor {
+                    def: sensor.def.clone(),
+                    state: sensor.state.clone(),
+                });
+                ids.push(sensor_id)
+            }
+
+            let component = SpComponent::try_from(cfg.id.as_str()).unwrap();
+            let prev = by_component.insert(component, ids);
+            assert!(prev.is_none(), "component ID collision!");
+        }
+        Self { sensors, by_component }
+    }
+
+    fn sensor_for_component<'sensors>(
+        &'sensors self,
+        component: &SpComponent,
+        index: BoundsChecked,
+    ) -> Option<&'sensors Sensor> {
+        let &id = self.by_component.get(component)?.get(index.0 as usize)?;
+        self.sensors.get(id as usize)
+    }
+
+    pub(crate) fn num_component_details(
+        &self,
+        component: &SpComponent,
+    ) -> Option<u32> {
+        let len = self
+            .by_component
+            .get(component)?
+            .len()
+            .try_into()
+            .expect("why would you have more than `u32::MAX` sensors?");
+        Some(len)
+    }
+
+    /// This method returns an `Option` because the component's details might
+    /// be a port status rather than a measurement, if we eventually decide to
+    /// implement port statuses in the simulated sidecar...
```

Comment:
I think it'd be fine for this to unwrap `sensor_for_component` and expect the caller to handle port statuses on its own and return a non-optional `ComponentDetails`. No objection to leaving this as-is either, though.

Reply:
My thinking was that it would be easier for the simulated sidecar to do that by writing something along the lines of
```rust
sensors
    .component_details(component)
    .unwrap_or_else(/* get port status details */)
```
but,  I suppose you could just look for port statuses first...

The other thing is, I don't actually know whether in real life, the same component will have both port statuses *and* sensor measurements or not? I can go check...

Reply:
Poking at a real sidecar, it appears that the  Monorail component, at least, only has port statuses...